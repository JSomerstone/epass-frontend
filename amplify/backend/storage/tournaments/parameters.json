{
    "tableName": "tournaments",
    "partitionKeyName": "year",
    "partitionKeyType": "N",
    "sortKeyName": "id",
    "sortKeyType": "S"
}