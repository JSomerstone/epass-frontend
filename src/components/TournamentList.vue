<template>
  <div class="tournament-list">
    <b-table
      :data="items"
      :bordered="false"
      :striped="true"
      :hoverable="true"
      :loading="isLoading"
      default-sort="name"
      default-sort-direction="asc"
      detailed
      detail-key="id"
      ref="table"
    >
      <template slot-scope="props">
        <b-table-column field="name" label="Name" sortable class="name-column">
          <a @click="toggle(props.row)">
            {{ props.row.name }}
          </a>
          <b-tooltip label="Tournament is locked" v-if="props.row.locked" type="is-info">
            <b-icon icon="lock" size="is-small"/>
          </b-tooltip>
        </b-table-column>
        <b-table-column field="dates" label="Date" sortable :custom-sort="sortDates" class="date-column">
          {{ props.row.dates | dateRange }} 
        </b-table-column>
        <b-table-column field="international" label="Type" sortable>
          {{ props.row.international ? "International" : "National" }}
        </b-table-column>
        <b-table-column field="city" label="Location" sortable>
            {{ props.row.city }}, {{ props.row.country }}
        </b-table-column>
        <b-table-column label="Statistics" title="Games / official duties">
            {{ getStatistics(props.row) }}
        </b-table-column>
      </template>
      <template slot="detail" slot-scope="props">
        <tournament-detail-row :tournament="props.row" />
      </template>
      <template slot="empty" v-if="!loading">
        <p>
          No tournaments to show, use the form above to add new or
          <b-button @click="handleReload" ref="reloadBtn">Reload</b-button>
        </p>
      </template>
    </b-table>
  </div>
</template>
<style>
.tournament-list {
  margin-bottom: 2em;
}
</style>
<script>
import TournamentDetailRow from "./TournamentDetailRow";

export default {
  components: {
    TournamentDetailRow
  },
  props: {
    loading: {
      type: Boolean,
      default: false
    }, 
    items: {
      type: Array,
      default: () => []
    },
    year: {
      type: [Number, String],
      required: true
    }
  },
  computed: {
    isEmpty: function() {
      return this.items.length == 0;
    },
    isLoading: function() {
      return this.$store.getters['tournaments/loading'];
    }
  },
  methods: {
    sortDates(rowA, rowB, isAsc) {
      let diff = new Date(rowA.dates[0]).getTime() 
        - new Date(rowB.dates[0]).getTime();
      return isAsc ? diff : diff * -1;
    },
    toggle(row) {
        this.$refs.table.toggleDetails(row)
    },
    handleReload: function() {
      this.$store.dispatch("tournaments/load", { year: this.year, force: true });
    },
    getStatistics(tournament) {
      const current = this.$store.getters['referees/current'];
      if (! current) {
        return;
      }
      const referee = tournament.referees.find( r => r.id == current.id );
      const { games= "-", tenSeconds= "-"} = referee || {};
      return `${games} / ${tenSeconds}`;
    }
  },
  filters: {
    dateRange: (dates) => {
      if (! dates) {
        return;
      } else if (dates[0] == dates[1]) {
        return new Date(dates[0]).toLocaleDateString();
      } else {
        return dates.map(
          d => new Date(d).toLocaleDateString()
        ).join(" - ");
      }
    }
  }
}
</script>