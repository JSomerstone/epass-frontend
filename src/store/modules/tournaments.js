import { listTournaments, getTournament } from "../../graphql/queries";
import { API } from "aws-amplify";
import { createTournament, updateTournament, createComment, deleteComment, deleteTournament } from '../../graphql/mutations';
import { notifyException, successMessage, warningMessage } from '../../utils/notificationUtils';


const state = {
  loading: false,
  tournaments: [],
  filteredTournaments: [],
  selectedTournaments: [],
  filter: { show: "own" },
  wip: null,
  showTournamentForm: false,
  teams: [],
  debug: false
};

const mutationTypes = {
  SET_LOADING: "set-loading",
  NEW_TOURNAMENT: "create-tournament",
  UPDATE_TOURNAMENT: "update-tournament",
  SET_TOURNAMENTS: "set-tournaments",
  SET_SELECTED_TOURNAMENTS: "set-selected-tournaments",
  SET_WIP: "set-wip",
  SET_SHOW_FORM: "set-show-form",
  SET_TEAMS: "set-teams",
  ADD_TEAM: "add-team",
  ADD_COMMENT: "add-comment",
  DELETE_COMMENT: "delete-comment",
  SET_FILTER: "set-filter",
  SET_FILTERED: "set-filtered",
};

const mutations = {
  [mutationTypes.SET_LOADING](state, loading) {
    state.loading = Boolean(loading);
  },
  [mutationTypes.NEW_TOURNAMENT](state, tournament) {
    state.tournaments.push(tournament);
  },
  [mutationTypes.UPDATE_TOURNAMENT](state, tournament) {
    const index = state.tournaments.findIndex(t => t.id === tournament.id);
    state.tournaments[index] = tournament;
  },
  [mutationTypes.SET_TOURNAMENTS](state, tournaments) {
    state.tournaments = tournaments;
  },
  [mutationTypes.SET_SELECTED_TOURNAMENTS](state, tournaments) {
    state.selectedTournaments = tournaments;
  },
  [mutationTypes.SET_FILTERED](state, tournaments) {
    state.filteredTournaments = tournaments;
  },
  [mutationTypes.SET_TEAMS](state, list) {
    state.teams = list;
  },
  [mutationTypes.ADD_TEAM](state, team) {
    state.teams.push(team);
    state.teams.sort();
    localStorage.setItem("teams", JSON.stringify(state.teams));
  },
  [mutationTypes.ADD_COMMENT](state, comment) {
    console.log({ wip: state.wip, comment });
  },
  [mutationTypes.DELETE_COMMENT](state, comment) {
    const index = state.wip.comments.findIndex(
      c => c.id == comment.id
    );
    state.wip.comments.splice(index, 1);
  },
  [mutationTypes.SET_WIP](state, tournament) {
    state.wip = tournament;
  },
  [mutationTypes.SET_SHOW_FORM](state, show) {
    state.showTournamentForm = Boolean(show);
  },
  [mutationTypes.SET_FILTER](state, show) {
    state.filter = show;
  },
};

const actions = {
  setLoading({ commit }, { loading }) {
    commit(mutationTypes.SET_LOADING, loading);
  },
  addComment: async ({ commit }, { comment, onSuccess = () => { } } ) => {
    API.graphql({
      query: createComment,
      variables: { input: comment },
    })
      .then(result => {
        commit(mutationTypes.ADD_COMMENT, result.data.createComment);
        console.log({ action: 'AddComment', ...result.data.createComment });
        onSuccess(result.data.createComment);
      })
      .catch(notifyException);
  },
  deleteComment: async ({ commit }, { id, onSuccess = () => { } }) => {
    commit(mutationTypes.SET_LOADING, true);
    API.graphql({
      query: deleteComment,
      variables: { input: { id } },
    })
      .then(result => {
        console.log({ action: 'RemoveComment', ...result.data });
        onSuccess(result);
      })
      .catch(notifyException)
      .finally(() => commit(mutationTypes.SET_LOADING, false));
  },
  load: async ({ commit, dispatch, state }, { year, force = false }) => {
    if (state.tournaments.length && state.tournaments[0].year == year && !force) {
      return;
    }
    commit(mutationTypes.SET_LOADING, true);
    try {
      const result = await API.graphql({
        query: listTournaments,
        variables: {
          filter: {
            year: { eq: year },
          },
        },
      });
      const tournaments = result.data.listTournaments.items;
      commit(mutationTypes.SET_TOURNAMENTS, tournaments);
      dispatch("filter");
    } catch (error) {
      notifyException(error);
    }
    commit(mutationTypes.SET_LOADING, false);
  },
  loadRefereesTournaments: async ({ commit }, { refereeId, year }) => {
    commit(mutationTypes.SET_LOADING, true);
    API.graphql({
      query: listTournaments,
      variables: {
        filter: { year: { eq: year } },
      },
    })
      .then(
        result => result.data.listTournaments.items.filter(
          t => t.td == refereeId || t.referees.find(r => r.id == refereeId)
        )
      )
      .then(
        tournaments => commit(mutationTypes.SET_SELECTED_TOURNAMENTS, tournaments)
    )
      .catch(notifyException)
      .finally(() => commit(mutationTypes.SET_LOADING, false));
  },
  loadTournament: ({ commit }, { id, onSuccess = () => {}, onFailure = () => {} }) => {
    commit(mutationTypes.SET_LOADING, true);
    API.graphql({
      query: getTournament,
      variables: { id },
    })
      .then((result) => {
        if (!result.data.getTournament) {
          throw { message: "Tournament not found!" };
        }
        commit(mutationTypes.SET_WIP, result.data.getTournament);
        onSuccess(result.data.getTournament);
      })
      .catch(e => {
        notifyException(e);
        onFailure();
      })
      .finally(() => {
        commit(mutationTypes.SET_LOADING, false);
      });
  },
  filter: ({ commit, rootGetters, state }) => {
    commit(mutationTypes.SET_LOADING, true);
    if (state.filter.show == "all") {
      commit(mutationTypes.SET_FILTERED, state.tournaments);
    } else {
      const { id = null } = rootGetters["referees/current"];
      const result = state.tournaments.filter((t) => {
        let found = t.referees.find((r) => r.id == id);
        return t.td == id || found;
      });
      commit(mutationTypes.SET_FILTERED, result);
    }
    commit(mutationTypes.SET_LOADING, false);
  },
  loadTeams: async ({ commit }) => {
    const result = JSON.parse(localStorage.getItem("teams") || "[]");
    result.sort();
    commit(mutationTypes.SET_TEAMS, result);
  },
  addTeam: async ({ commit, state }, { team }) => {
    const existing = Boolean(
      state.teams.find(old => old.toLowerCase() == team.toLowerCase()),
    );
    if (!existing) {
      commit(mutationTypes.ADD_TEAM, team);
    }
  },
  create: async ({ commit, dispatch, rootGetters }, { tournament, onSuccess = () => { } }) => {
    commit(mutationTypes.SET_LOADING, true);
    try {
      tournament.createdBy = rootGetters["referees/current"].id || null;
      const { comments } = tournament;
      delete tournament.comments;
      const result = await API.graphql({
        query: createTournament,
        variables: { input: tournament.toJson() },
      });
      const { id } = result.data.createTournament;
      commit(mutationTypes.NEW_TOURNAMENT, result.data.createTournament);
      successMessage("Tournament saved");
      dispatch("load", { year: tournament.year, force: true });
      comments.items.forEach(comment => {
        dispatch("addComment", {
          comment: { ...comment, commentTournamentId: id }
        });
      });
      onSuccess(result.data.createTournament);
    } catch (err) {
      notifyException(err);
    }
    commit(mutationTypes.SET_LOADING, false);
  },
  update: async ({ commit, dispatch , rootGetters}, { tournament, onSuccess = () => { } }) => {
    commit(mutationTypes.SET_LOADING, true);
    try {
      if (!tournament.createdBy) {
        tournament.createdBy = rootGetters["referees/current"].id || null;
      }
      const result = await API.graphql({
        query: updateTournament,
        variables: { input: tournament.toJson() },
      });
      commit(mutationTypes.UPDATE_TOURNAMENT, result.data.updateTournament);
      commit(mutationTypes.SET_WIP, null);
      dispatch("load", { year: tournament.year, force: true });
      successMessage("Tournament updated");
      onSuccess(result.data.updateTournament);
    } catch (err) {
      notifyException(err);
    }
    commit(mutationTypes.SET_LOADING, false);
  },
  delete: async ({ commit, dispatch, rootGetters }, { tournament, onSuccess = () => { } }) => {
    const { id, year, createdBy, td } = tournament;
    const refereeId = rootGetters["referees/current"].id || false;
    if (createdBy !== refereeId && td !== refereeId) {
      warningMessage("Only original creator or TD can delete a tournament");
      return;
    }
    commit(mutationTypes.SET_LOADING, true);
    API.graphql({
      query: deleteTournament,
      variables: { input: { id } },
    }).then(
      result => {
        dispatch("load", { year, force: true });
        onSuccess(result);
      }
    )
      .catch(notifyException)
      .finally(() => commit(mutationTypes.SET_LOADING, false));
  },
  setFilter: ({ commit, dispatch }, { filter }) => {
    commit(mutationTypes.SET_FILTER, filter);
    dispatch("filter");
  },
  reset: ({ commit }) => {
    commit(mutationTypes.SET_TOURNAMENTS, []);
    commit(mutationTypes.SET_FILTERED, []);
    commit(mutationTypes.SET_WIP, {});
  },
};

const tournaments = {
  namespaced: true,
  state,
  getters: {
    loading: state => state.loading,
    all: state => state.filteredTournaments,
    refereesTournaments: state => state.selectedTournaments,
    byId: state => id => state.tournaments.find( t => t.id == id ),
    teams: state => state.teams,
    showForm: state => state.showTournamentForm,
    wip: state => state.wip,
    hasWip: state => Boolean(state.wip),
    filter: state => state.filter,
    debug: state => state.debug,
  },
  mutations,
  actions,
};
export default tournaments;
