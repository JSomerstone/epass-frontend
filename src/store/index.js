import Vue from "vue";
import Vuex from "vuex";

import axios from "axios";
import config from "../config/index";

import tournaments from "./modules/tournaments";
import referees from "./modules/referees";
import countries from "./modules/countries";
import auth from "./modules/auth";

axios.defaults.baseURL = config.backendUrl;
Vue.use(Vuex);

const store = new Vuex.Store({
  modules: {
    tournaments,
    referees,
    countries,
    auth
  }
});

export default store;