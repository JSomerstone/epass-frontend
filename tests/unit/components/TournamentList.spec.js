import { mount, RouterLinkStub } from "@vue/test-utils";
import Vuex from "vuex";
import localVue from "../localVue";
import store from "../mockStore";
import TournamentList from "@/components/TournamentList";
import tournaments from "../../data/tournaments.json"

describe("TournamentList", () => {

  beforeEach(() => {
  });

  it("Renders correctly", () => {
    const wrapper = mount(TournamentList, {
      localVue,
      store: new Vuex.Store(store),
      propsData: {
        items: tournaments[2020],
        year: 2020
      }
    });
    expect(wrapper).toMatchSnapshot();
  });

  it("Renders correctly with empty list", () => {
    const wrapper = mount(TournamentList, {
      localVue,
      store: new Vuex.Store(store),
      propsData: {
        year: 2020
      }
    });
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.vm.isEmpty).toBe(true);
  });

  it("Reloading dispatches tournaments/load", async () => {
    const wrapper = mount(TournamentList, {
      localVue,
      store: new Vuex.Store(store),
      propsData: {
        year: 2020
      }
    });
    const { load } = store.modules.tournaments.actions;
    await wrapper.findComponent({ ref: "reloadBtn" }).trigger("click");
    
    expect(load).toHaveBeenCalled();
  });

  it("Clicking tournament name opens details", async () => {
    const wrapper = mount(TournamentList, {
      localVue,
      store: new Vuex.Store(store),
      stubs: {  RouterLink: RouterLinkStub },
      propsData: {
        items: tournaments[2020],
        year: 2020
      }
    });
    await wrapper.find("td a").trigger("click");

    expect(wrapper).toMatchSnapshot();
    await wrapper.vm.toggle(0);
  });

  it("Tournaments can be sorted by date", async () => {
    const wrapper = mount(TournamentList, {
      localVue,
      store: new Vuex.Store(store),
      propsData: {
        items: tournaments[2020],
        year: 2020
      }
    });
    expect(wrapper.find("td.date-column").text()).toBe("2/20/2020 - 2/22/2020");
    await wrapper.find("td.date-column").trigger("click");
    expect(wrapper.find("td.date-column").text()).toBe("3/20/2020");

    const row0 = { dates: ["2020-08-25 00:00:00", "2020-08-26 00:00:00"] };
    const row1 = { dates: ["2020-08-24 00:00:00", "2020-08-24 00:00:00"] };
    expect(wrapper.vm.sortDates(row0, row1, true)).toBe(24 * 60 * 60 * 1000);
  });
});