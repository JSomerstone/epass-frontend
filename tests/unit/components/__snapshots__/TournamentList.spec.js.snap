// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TournamentList Clicking tournament name opens details 1`] = `
<div class="tournament-list">
  <div class="b-table">
    <div class="field table-mobile-sort">
      <div class="field has-addons">
        <div class="control is-expanded"><span class="select is-fullwidth"><select><!----> <!----> <option value="[object Object]">
                    Name
                </option><option value="[object Object]">
                    Date
                </option><option value="[object Object]">
                    Type
                </option><option value="[object Object]">
                    Location
                </option><!----></select></span>
          <!---->
        </div>
        <div class="control"><button class="button is-primary"><span class="icon is-small"><i class="mdi mdi-arrow-up"></i></span></button></div>
      </div>
    </div>
    <!---->
    <div class="table-wrapper has-mobile-cards">
      <table class="table is-striped is-hoverable">
        <thead>
          <tr>
            <th width="40px"></th>
            <!---->
            <th class="is-current-sort is-sortable">
              <div class="th-wrap">Name <span class="icon is-small"><i class="mdi mdi-arrow-up"></i></span></div>
            </th>
            <th class="is-sortable">
              <div class="th-wrap">Date <span class="icon is-small is-invisible"><i class="mdi mdi-arrow-up"></i></span></div>
            </th>
            <th class="is-sortable">
              <div class="th-wrap">Type <span class="icon is-small is-invisible"><i class="mdi mdi-arrow-up"></i></span></div>
            </th>
            <th class="is-sortable">
              <div class="th-wrap">Location <span class="icon is-small is-invisible"><i class="mdi mdi-arrow-up"></i></span></div>
            </th>
            <th class="">
              <div class="th-wrap">Statistics
                <!---->
              </div>
            </th>
            <!---->
          </tr>
          <!---->
          <!---->
        </thead>
        <tbody>
          <tr draggable="false" class="">
            <td class="chevron-cell"><a role="button"><span class="icon"><i class="mdi mdi-chevron-right mdi-24px"></i></span></a></td>
            <!---->
            <td data-label="Name" class="name-column"><a>
                Bogus
              </a> <span data-label="Tournament is locked" class="is-info is-top is-medium b-tooltip" style="transition-delay: 0ms;"><span class="icon is-small"><i class="mdi mdi-lock"></i></span></span></td>
            <td data-label="Date" class="date-column">
              3/20/2020
            </td>
            <td data-label="Type" class="">
              National
            </td>
            <td data-label="Location" class="">
              Helsinki, Finland
            </td>
            <td data-label="Statistics" class="" title="Games / official duties">
              5 / 6
            </td>
            <!---->
          </tr>
          <!---->
          <!---->
          <tr draggable="false" class="">
            <td class="chevron-cell"><a role="button"><span class="icon is-expanded"><i class="mdi mdi-chevron-right mdi-24px"></i></span></a></td>
            <!---->
            <td data-label="Name" class="name-column"><a>
                Example Tournament 2020
              </a>
              <!---->
            </td>
            <td data-label="Date" class="date-column">
              2/20/2020 - 2/22/2020
            </td>
            <td data-label="Type" class="">
              International
            </td>
            <td data-label="Location" class="">
              Malmö, Sweden
            </td>
            <td data-label="Statistics" class="" title="Games / official duties">
              6 / 5
            </td>
            <!---->
          </tr>
          <tr class="detail">
            <td colspan="6">
              <div class="detail-container">
                <div class="columns">
                  <div class="column is-one-third"><a><button type="button" outline="" class="button is-info"><span class="icon is-small"><i class="mdi mdi-folder-open-outline"></i></span> <span>
        Open
      </span>
                        <!----></button></a>
                    <div><strong>Your statistics:</strong>
                      <ul>
                        <li>
                          6
                          Games as referee
                        </li>
                        <li>
                          5
                          Games
                          as table official / 10 sec timer
                        </li>
                      </ul>
                    </div>
                  </div>
                  <div class="column is-one-third">
                    <div><strong>TD:</strong> John Doe [USA]
                    </div>
                    <div><strong>Referees:</strong>
                      <ul>
                        <li>
                          Jane Doe [CAN]
                        </li>
                        <li>
                          John Doe [USA]
                        </li>
                        <li>
                          Noob Nederson [SWE]
                        </li>
                      </ul>
                    </div>
                  </div>
                  <div class="column is-one-third"><strong>Teams</strong>
                    <ul>
                      <li>
                        Sweden
                      </li>
                      <li>
                        Finland
                      </li>
                      <li>
                        Denmark
                      </li>
                    </ul>
                  </div>
                </div>
              </div>
            </td>
          </tr>
          <!---->
        </tbody>
        <!---->
      </table>
    </div>
    <!---->
  </div>
</div>
`;

exports[`TournamentList Renders correctly 1`] = `
<div class="tournament-list">
  <div class="b-table">
    <!---->
    <!---->
    <div class="table-wrapper has-mobile-cards">
      <table class="table is-striped is-hoverable">
        <!---->
        <tbody>
          <tr draggable="false" class="">
            <td class="chevron-cell"><a role="button"><span class="icon"><i class="mdi mdi-chevron-right mdi-24px"></i></span></a></td>
            <!---->
            <td data-label="Name" class="name-column"><a>
                Example Tournament 2020
              </a>
              <!---->
            </td>
            <td data-label="Date" class="date-column">
              2/20/2020 - 2/22/2020
            </td>
            <td data-label="Type" class="">
              International
            </td>
            <td data-label="Location" class="">
              Malmö, Sweden
            </td>
            <td data-label="Statistics" class="" title="Games / official duties">
              6 / 5
            </td>
            <!---->
          </tr>
          <!---->
          <!---->
          <tr draggable="false" class="">
            <td class="chevron-cell"><a role="button"><span class="icon"><i class="mdi mdi-chevron-right mdi-24px"></i></span></a></td>
            <!---->
            <td data-label="Name" class="name-column"><a>
                Bogus
              </a> <span data-label="Tournament is locked" class="is-info is-top is-medium b-tooltip" style="transition-delay: 0ms;"><span class="icon is-small"><i class="mdi mdi-lock"></i></span></span></td>
            <td data-label="Date" class="date-column">
              3/20/2020
            </td>
            <td data-label="Type" class="">
              National
            </td>
            <td data-label="Location" class="">
              Helsinki, Finland
            </td>
            <td data-label="Statistics" class="" title="Games / official duties">
              5 / 6
            </td>
            <!---->
          </tr>
          <!---->
          <!---->
        </tbody>
        <!---->
      </table>
    </div>
    <!---->
  </div>
</div>
`;

exports[`TournamentList Renders correctly with empty list 1`] = `
<div class="tournament-list">
  <div class="b-table">
    <!---->
    <!---->
    <div class="table-wrapper has-mobile-cards">
      <table class="table is-striped">
        <!---->
        <tbody>
          <tr class="is-empty">
            <td colspan="1">
              <p>
                No tournaments to show, use the form above to add new or
                <button type="button" class="button">
                  <!----> <span>Reload</span>
                  <!----></button></p>
            </td>
          </tr>
        </tbody>
        <!---->
      </table>
    </div>
    <!---->
  </div>
</div>
`;
